### DISCUSSION TITLE
State (back button restoral of state, scroll, ...)

### DISCUSSION THREAD

**Discussion by apatrida** [varabyte/kobweb#552] [General]
Hi, 

I do not see clearly in the docs any management by the router of state restoral, scroll position restoral.  Is this because state is intended to be managed at the app level, via the URL, or some other way when navigating back through routes previously visited?

For example, the playground app, if I enter a user name then go to another tab, and return, the user name is lost.  Obviously it isn't intending to handle that in this codebase, but probably should.  How would it?

**Comment by bitspittle** [comment]
Saved state on web is supported by local storage and session storage concepts. They are the same API but local storage exists on disk essentially until the user clears it while session storage is tied to the current tab and will be cleared when it is closed.

You can read more about it here: https://developer.mozilla.org/en-US/docs/Web/API/Web_Storage_API

If you run `kobweb create app` you can see [usage of local storage here](https://github.com/varabyte/kobweb-templates/blob/7b676edb93160d789c1a88fa01142358739e95cc/app/default/site/src/site/AppEntry.kt.ftl#L23).

Scroll state is handled by the browser -- I'm not really sure I have much control over that. That said, you can use fragments (using the '#' character) as a way to control scrolling. For example, something like https://bitspittle.dev/blog/2022/kotlin-site#conclusion

**Reply by apatrida** [reply]
Any router hooks for using push state on history change, or just the browser directly to try and intercept there?  (but then without knowledge of routes, router, ...)

**Reply by bitspittle** [reply]
[There is a method provided for intercepting routes](https://github.com/varabyte/kobweb/blob/e4984af8ae9a4141bb7c9f965d338ddcbf4f6fc9/frontend/kobweb-core/src/jsMain/kotlin/com/varabyte/kobweb/navigation/Router.kt#L297), but I'm not sure it's what you're looking for (this is more a way to handle updating the final URL and less about being a hook during which you update relevant state)

**Reply by bitspittle** [reply]
If you have a specific use case that you're struggling with, you can mention it here and I can let you know if there's an existing way to accomplish this vs. maybe Kobweb needs to add a new feature. At the moment I can't tell.
